src/appsync.ts

```diff
@@ -152,419 +164,447 @@ export class AppsyncVTL extends VTL {
-      | "returnType"
-      | keyof ReturnType<
-          AppsyncResolver<Arguments, Result, Source>["getResolvableFields"]
-        >
-    >
-  ): appsync.ResolvableField {
-    const fields = this.getResolvableFields(api);
-
-    return new appsync.ResolvableField({
+  readonly api: appsync.GraphqlApi;
+}
+
+/**
+ * Generate a resolvable field to use with AppSync CDK's Code field strategy.
+ *
+ * ```ts
+ * import * as appsync from "@aws-cdk/aws-appsync-alpha";
+ *
+ * const api = new appsync.GraphQLApi(..);
+ *
+ * ```ts
+ * const getPersonField = new AppsyncField<{id: string}, Person | undefined>(
+ *   {
+ *     api,
+ *     returnType: appsync.Field.string(),
+ *     args: {
+ *       argName: appsync.Field.string()
+ *     }
+ *   },
+ *   ($context, id) => {
+ *     const person = table.get({
+ *       key: {
+ *         id: $util.toDynamoDB(id)
+ *       }
+ *     });
+ *     return person;
+ *   });
+ * ```
+ *
+ * // code first person type
+ * const personType = api.addType(appsync.ObjectType(...));
+ *
+ * api.addQuery("getPerson", getPersonField)
+ * ```
+ *
+ * @param api app sync API which the data sources construct will be added under.
+ * @param returnType the code-first graphql return type of this field
+ * @param options optional fields like arguments and directive to add to the schema
+ */
+export class AppsyncField<
+  Arguments extends ResolverArguments,
+  Result,
+  Source = undefined
+> extends appsync.ResolvableField {
+  /**
+   * This static property identifies this class as an AppsyncResolver to the TypeScript plugin.
+   */
+  public static readonly FunctionlessType = "AppsyncField";
+
+  constructor(
+    options: AppsyncFieldOptions,
+    resolve: ResolverFunction<Arguments, Result, Source>
+  ) {
+    const fields = synthResolvers(
+      options.api,
+      validateFunctionDecl(resolve, "AppsyncField")
+    );
+    super({
       ...options,
-      returnType,
       dataSource: fields.dataSource,
       pipelineConfig: fields.pipelineConfig,
       requestMappingTemplate: fields.requestMappingTemplate,
       responseMappingTemplate: fields.responseMappingTemplate,
     });
   }
+}
 
-  private getResolvableFields(api: appsync.GraphqlApi) {
-    const resolverCount = countResolvers(this.decl);
+function synthResolvers(api: appsync.GraphqlApi, decl: FunctionDecl) {
+  const resolverCount = countResolvers(decl);
 
-    const [pipelineConfig, responseMappingTemplate, innerTemplates] =
-      synthesizeFunctions(api, this.decl);
+  const [pipelineConfig, responseMappingTemplate, innerTemplates] =
+    synthesizeFunctions(api, decl);
 
-    // mock integration
-    if (resolverCount === 0) {
-      if (pipelineConfig.length !== 0) {
-        throw new Error(
-          `expected 0 functions in pipelineConfig, but found ${pipelineConfig.length}`
-        );
-      }
+  // mock integration
+  if (resolverCount === 0) {
+    if (pipelineConfig.length !== 0) {
+      throw new Error(
```

Comment:
In my PR I removed this check. I am not sure I understand the purpose, do we think we'll mess up? Couldn't an integration provide only vtl in the future?

Reply:
Yeah, I don't really know what this is checking for either. Seems to check for a situation where we rendered a pipeline without ever calling a resolver? Seems like internal bug.

Reply:
Which PR should we merge first?